// lamma-ui/lib/divider/divider.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = Divider;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * @author sunyue
 *
 * 请在这里填写当组件 **Divider** 的描述信息，可以使用 **Markdown** 语法。
 */
function Divider(props) {
    var _classNames;

    var prefixCls = 'rc';

    var type = props.type,
        className = props.className,
        children = props.children,
        dashed = props.dashed,
        restProps = _objectWithoutProperties(props, ['type', 'className', 'children', 'dashed']);

    var classString = (0, _classnames2.default)(className, prefixCls + '-divider', prefixCls + '-divider-' + type, (_classNames = {}, _classNames[prefixCls + '-divider-with-text\''] = children, _classNames[prefixCls + '-divider-dashed'] = !!dashed, _classNames));
    return _react2.default.createElement(
        'div',
        _extends({ className: classString }, restProps),
        children && _react2.default.createElement(
            'span',
            { className: prefixCls + '-divider-inner-text' },
            children
        )
    );
}

Divider.propTypes = {
    /**
     * 请在这里填写组件当前属性的描述信息
     */
    name: _propTypes2.default.string,
    type: _propTypes2.default.oneOf(['horizontal', 'vertical']),
    className: _propTypes2.default.string,
    children: _propTypes2.default.string,
    dashed: _propTypes2.default.bool
};

Divider.defaultProps = {
    name: '',
    type: 'horizontal',
    className: '',
    children: '',
    dashed: false
};
