// lamma-ui/lib/input-number/input-number.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = InputNumber;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _rcInputNumber = require('rc-input-number');

var _rcInputNumber2 = _interopRequireDefault(_rcInputNumber);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var prefixIN = 'rc-input-number';
/**
 * @author sunyue
 * 请在这里填写当组件 **InputNumber** 的描述信息，可以使用 **Markdown** 语法。
 */
function InputNumber(props) {
    var _classNames;

    var className = props.className,
        size = props.size,
        other = _objectWithoutProperties(props, ['className', 'size']);

    var inputNumberClass = (0, _classnames2.default)((_classNames = {}, _classNames[prefixIN + '-lg'] = size === 'large', _classNames[prefixIN + '-sm'] = size === 'small', _classNames[className] = !!className, _classNames));
    return _react2.default.createElement(_rcInputNumber2.default, _extends({ className: inputNumberClass }, other));
}

InputNumber.propTypes = {
    /**
    * 输入框大小
    */
    size: _propTypes2.default.oneOf(['large', 'default', 'small']),
    /**
    * 数字输入框样式
    */
    className: _propTypes2.default.string,
    /**
    * 最小值
    */
    min: _propTypes2.default.number,
    /**
    * 最大值
    */
    max: _propTypes2.default.number,
    /**
    * 步长，可改为小数
    */
    step: _propTypes2.default.number,
    /**
    * 变化回调
    */
    onChange: _propTypes2.default.func,
    /**
    * 是否禁用
    */
    disabled: _propTypes2.default.bool
};

InputNumber.defaultProps = {
    size: 'default',
    className: '',
    min: -Infinity,
    max: Infinity,
    step: 1,
    disabled: false,
    onChange: function onChange() {}
};
