// lamma-ui/lib/calendar/calendar.js
'use strict';

exports.__esModule = true;
exports.zhCN = exports.enUS = exports.Picker = exports.FullCalendar = exports.MonthCalendar = exports.RangeCalendar = undefined;
exports.default = Calendar;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _rcCalendar = require('rc-calendar');

var _rcCalendar2 = _interopRequireDefault(_rcCalendar);

var _RangeCalendar2 = require('rc-calendar/lib/RangeCalendar');

var _RangeCalendar3 = _interopRequireDefault(_RangeCalendar2);

var _MonthCalendar2 = require('rc-calendar/lib/MonthCalendar');

var _MonthCalendar3 = _interopRequireDefault(_MonthCalendar2);

var _FullCalendar2 = require('rc-calendar/lib/FullCalendar');

var _FullCalendar3 = _interopRequireDefault(_FullCalendar2);

var _Picker2 = require('rc-calendar/lib/Picker');

var _Picker3 = _interopRequireDefault(_Picker2);

var _en_US = require('rc-calendar/lib/locale/en_US');

var _en_US2 = _interopRequireDefault(_en_US);

var _zh_CN = require('rc-calendar/lib/locale/zh_CN');

var _zh_CN2 = _interopRequireDefault(_zh_CN);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @author [SunLei](http://git.4paradigm.com/u/sunlei)
 * 
 * 详细文档请参考：http://react-component.github.io/calendar
 * 
 * 大部分场景下请使用 [DateTimePicker](datetime-picker.html) 组件进行日期、时间的选取
 */
function Calendar(props) {
    return _react2.default.createElement(_rcCalendar2.default, props);
}

exports.RangeCalendar = _RangeCalendar3.default;
exports.MonthCalendar = _MonthCalendar3.default;
exports.FullCalendar = _FullCalendar3.default;
exports.Picker = _Picker3.default;
exports.enUS = _en_US2.default;
exports.zhCN = _zh_CN2.default;
