// lamma-ui/lib/tabs/tabs.js
'use strict';

exports.__esModule = true;
exports.SwipeableTabContent = exports.TabContent = exports.SwipeableInkTabBar = exports.ScrollableInkTabBar = exports.ScrollableTabBar = exports.InkTabBar = exports.TabBar = exports.TabPane = undefined;

var _rcTabs = require('rc-tabs');

Object.defineProperty(exports, 'TabPane', {
    enumerable: true,
    get: function get() {
        return _rcTabs.TabPane;
    }
});

var _TabBar = require('rc-tabs/lib/TabBar');

Object.defineProperty(exports, 'TabBar', {
    enumerable: true,
    get: function get() {
        return _interopRequireDefault(_TabBar).default;
    }
});

var _InkTabBar = require('rc-tabs/lib/InkTabBar');

Object.defineProperty(exports, 'InkTabBar', {
    enumerable: true,
    get: function get() {
        return _interopRequireDefault(_InkTabBar).default;
    }
});

var _ScrollableTabBar = require('rc-tabs/lib/ScrollableTabBar');

Object.defineProperty(exports, 'ScrollableTabBar', {
    enumerable: true,
    get: function get() {
        return _interopRequireDefault(_ScrollableTabBar).default;
    }
});

var _ScrollableInkTabBar = require('rc-tabs/lib/ScrollableInkTabBar');

Object.defineProperty(exports, 'ScrollableInkTabBar', {
    enumerable: true,
    get: function get() {
        return _interopRequireDefault(_ScrollableInkTabBar).default;
    }
});

var _SwipeableInkTabBar = require('rc-tabs/lib/SwipeableInkTabBar');

Object.defineProperty(exports, 'SwipeableInkTabBar', {
    enumerable: true,
    get: function get() {
        return _interopRequireDefault(_SwipeableInkTabBar).default;
    }
});

var _TabContent = require('rc-tabs/lib/TabContent');

Object.defineProperty(exports, 'TabContent', {
    enumerable: true,
    get: function get() {
        return _interopRequireDefault(_TabContent).default;
    }
});

var _SwipeableTabContent = require('rc-tabs/lib/SwipeableTabContent');

Object.defineProperty(exports, 'SwipeableTabContent', {
    enumerable: true,
    get: function get() {
        return _interopRequireDefault(_SwipeableTabContent).default;
    }
});
exports.default = Tabs;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _rcTabs2 = _interopRequireDefault(_rcTabs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @author sunlei
 * http://react-component.github.io/tabs/
 */
function Tabs(props) {
    return _react2.default.createElement(_rcTabs2.default, props);
}
